src/analysis_and_optimization/Mem_pattern.ml

```diff
@@ -0,0 +1,439 @@
+            pattern=
+              FunApp
+                ( CompilerInternal (FnReadParam (constrain_op, mem_pattern))
+                , List.map ~f:mod_expr args ) } )
+  | Assignment (((name : string), (ut : UnsizedType.t), lhs), rhs) ->
+      Assignment
+        ((name, ut, List.map ~f:(Index.map mod_expr) lhs), mod_expr rhs)
+  | IfElse (predicate, true_stmt, op_false_stmt) ->
+      IfElse
+        ( mod_expr predicate
+        , mod_stmt true_stmt
+        , Option.map ~f:mod_stmt op_false_stmt )
+  | Block stmts -> Block (List.map ~f:mod_stmt stmts)
+  | SList stmts -> SList (List.map ~f:mod_stmt stmts)
+  | For ({lower; upper; body; _} as loop) ->
+      Stmt.Fixed.Pattern.For
+        { loop with
+          lower= mod_expr lower; upper= mod_expr upper; body= mod_stmt body }
+  | TargetPE expr -> TargetPE (mod_expr expr)
+  | Return optional_expr -> Return (Option.map ~f:mod_expr optional_expr)
+  | Profile ((p_name : string), stmt) ->
+      Profile (p_name, List.map ~f:mod_stmt stmt)
+  | While (predicate, body) -> While (mod_expr predicate, mod_stmt body)
+  | Skip | Break | Continue | Decl _ -> pattern
+
+(**
+ * Modify statement patterns in the MIR from AoS <-> SoA and vice versa 
+ * @param mem_pattern A mem_pattern to modify expressions to. For the 
+ *  given memory pattern, this modifies 
+ *  statement patterns and expressions to it.
+ * @param stmt The statement to modify. 
+ * @param modifiable_set The name of the variable we are searching for.
+ *)
+and modify_stmt (mem_pattern : Common.Helpers.mem_pattern)
+    (Stmt.Fixed.({pattern; _}) as stmt) (modifiable_set : string Set.Poly.t) =
+  {stmt with pattern= modify_stmt_pattern mem_pattern pattern modifiable_set}
+
+(**
+ * Query an expression to check if any of it's named types exists in a set.
+ *  Return true if any of the names in `name_set` are in the expression.
+ **)
+let check_eigen_names (name_set : string Set.Poly.t)
+    (expr : Typed.Meta.t Expr.Fixed.t) : bool =
+  Set.Poly.is_empty (Set.inter name_set (query_eigen_names expr))
+
+(* Look through a statement to see whether it needs modified from
+ * SoA to AoS. Returns the set of object names that need demoted
+ * in a statement, if any.
+ *)
+let query_demotable_stmt (aos_exits : string Set.Poly.t)
+    (pattern : (Typed.t, int) Stmt.Fixed.Pattern.t) : string Set.Poly.t =
+  match pattern with
+  | Stmt.Fixed.Pattern.Assignment
+      ( ( (assign_name : string)
+        , (_ : UnsizedType.t)
+        , (_ : Expr.Typed.t Index.t list) )
+      , rhs ) ->
+      let all_rhs_eigen_names = query_eigen_names rhs in
+      let rhs_assign_type_non_eigen =
+        UnsizedType.contains_eigen_type (Expr.Typed.type_of rhs)
+      in
+      let rhs_fails =
+        match
+          Set.Poly.mem aos_exits assign_name && rhs_assign_type_non_eigen
+        with
+        | true -> all_rhs_eigen_names
+        | false -> Set.Poly.empty
+      in
+      let lhs_fails =
+        match check_eigen_names aos_exits rhs with
```

Comment:
So IIUC, this says that if any AoS variable is used in the RHS expr of an assignment, then the LHS should be AoS (and the RHS vars can probably be added here too if the LHS is being made AoS). But if the RHS is e.g. `f(x, y)`, with `f` being an SoA-able function, don't both `x` and `y` need to be AoS to demote the whole expression, rather than either one?

Another problem might be that if the RHS is actually using a non-eigen function of an AoS variable, like: `f(x, size(y))`, then the mem pattern of `y` shouldn't affect the mem pattern of the whole expression

Not sure if those are real problems, but might be worth including those examples in your tests if they aren't already

Reply:
> So IIUC, this says that if any AoS variable is used in the RHS expr of an assignment, then the LHS should be AoS (and the RHS vars can probably be added here too if the LHS is being made AoS). But if the RHS is e.g. f(x, y), with f being an SoA-able function, don't both x and y need to be AoS to demote the whole expression, rather than either one?

I tried cleaning up the logic a bit on this (it's under `check_names`). I think I had this wrong before, but now it should be setup so that as long as one of eigen expressions the RHS is SoA then the statement will continue to be SoA, but if all of the eigen expressions on the RHS are AoS then it will add the LHS assignee to the set of objects that need to be demoted to AoS

> Another problem might be that if the RHS is actually using a non-eigen function of an AoS variable, like: f(x, size(y)), then the mem pattern of y shouldn't affect the mem pattern of the whole expression

Indeed! Yes so I fixed this up as well and added an example that should capture this now.